pallets/nft-marketplace/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+		FailedNftTransfer,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add the given NFT to the gallery, putting it for sale.
+		///
+		/// Fails if
+		///   - the NFT is not found in [T::NonFungibles]
+		///   - `origin` is not the owner of the nft
+		///   - the nft is already for sale in the gallery
+		///   - transferring ownership of the NFT to this pallet's account fails
+		#[pallet::weight(10_000_000)]
+		pub fn add(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+			currency: CurrencyOf<T>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let seller = ensure_signed(origin.clone())?;
+
+			// Check that the seller is the owner of the nft
+			ensure!(
+				Self::is_owner(seller.clone(), class_id, instance_id)?,
+				Error::<T>::NotOwner,
+			);
+
+			// Check that the nft is not yet for sale
+			ensure!(
+				!<Gallery<T>>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyForSale
+			);
+
+			// Transfer the NFT to the parachain account
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &Self::account())
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Put the nft for sale
+			let sale = Sale {
+				seller,
+				price: AskingPrice { currency, amount },
+			};
+			<Gallery<T>>::insert(class_id, instance_id, sale.clone());
+			Self::deposit_event(Event::ForSale(sale));
+
+			Ok(())
+		}
+
+		/// Remove an NFT from sale
+		///
+		/// The sellers of an NFT that is for sale can call this extrinsic to reclaim
+		/// ownership over their NFT and remove it from sale.
+		///
+		/// Fails if
+		///   - the nft is not for sale
+		///   - `origin` is not the seller of the NFT
+		///   - transferring the ownership of the NFT back to the seller fails
+		#[pallet::weight(10_000_000)]
+		pub fn remove(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(who == sale.seller, Error::<T>::NotOwner);
+
+			// Transfer the NFT back to the seller, i.e., the original owner of this NFT
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &sale.seller)
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Remove the NFT from the gallery
+			<Gallery<T>>::remove(class_id, instance_id);
+			Self::deposit_event(Event::Removed {
+				class_id,
+				instance_id,
+			});
+
+			Ok(())
+		}
+
+		/// Buy the given nft
+		///
+		/// Fails if
+		///   - the NFT is not for sale
+		///   - `origin` is the seller of the NFT
+		///   - `origin` does not have enough balance of the currency the nft is being sold in
+		///   - transferring the asking price from the buyer to the seller fails
+		///   - transferring the nft to the buyer fails
+		#[pallet::weight(10_000_000)]
+		#[transactional]
+		pub fn buy(
```

Comment:
Either the (maximum) buy price needs to be included here, or listing should be updated to have some sort of unique ID.

Right now there's a "fun" scam you can pull with this pallet

1. List NFT for low price
2. Somebody guys to buy
3. Scammer removes then re-lists at a higher price before the buy transaction is included in a block
4. User gets charged a much higher price than planned

Anyone operating a full node that accepts transactions (the various providers for polkadot.js.org) as well as collators have the knowledge to do this sort of front-running.

Reply:
The best fix here is probably just to include a `max_price` in this call, and error if `max_price` is less than `sale_price`.

That way if there is a bit of a race where the seller drops the price just as someone is buying, the buy transaction will still succeed (for the lower price, even!).

If a buyer sets a higher `max_price` than asking and the seller is in a position to increase their price to match, well, at that point it's not really an attack anymore :)

Reply:
This is so cool, thanks a lot @branan for spotting this and proposing solutions 💯  I went ahead with the max_offer solution since it requires no extra data to be stored on chain.